Clear UR bar marks after seeking

sh0ckR6 opened the pull request:
Fixes #14470 

* Supports both seeking with keybinds and with the mouse
* Clears the UR bar after seeking (as suggested by [this comment](https://github.com/ppy/osu/issues/14470#issuecomment-904279618))

I also had to expose the `SongProgressBar` in `SongProgress` as a read-only property to avoid `Drawable` hierarchy calls but can be easily refactored to use the hierarchy instead.

sh0ckR6 committed:
Clear UR bar display on keyboard input

sh0ckR6 committed:
Clear UR bar display on seek with mouse

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs:
```diff
@@ -141,6 +144,10 @@ protected override void LoadComplete()
             arrow.Alpha = 0;
             arrow.Delay(200).FadeInFromZero(600);
+
+            var progressBar = Parent.ChildrenOfType<SongProgress>().FirstOrDefault();
```
`ChildrenOfType` is *never* used outside of testing methods. Also we don't do cross-component dependencies like this. `SongProgress` is not responsible for seeking, so shouldn't be the event source for other components to read from.

You should be resolving `GameplayClock` or `Player` or something and working from there.

sh0ckR6 replied:
Resolved in latest commit

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs:
```diff
@@ -10,15 +10,18 @@
 namespace osu.Game.Screens.Play.HUD.HitErrorMeters
 {
-    public class BarHitErrorMeter : HitErrorMeter
+    public class BarHitErrorMeter : HitErrorMeter, IKeyBindingHandler<GlobalAction>
```
The hit error meter should not be handling actions full stop.

sh0ckR6 replied:
Resolved in latest commit

peppy commented:
@sh0ckR6 just a bump on this, are you willing to apply the review feedback, or need any direction?

sh0ckR6 commented:
Things have been crazy all week at school, so I haven't really had time to work on the requested changes. Should've updated about this earlier in the week.

I'm free to work on this PR now, but I am a little confused about how resolving works.
Where does `[Resolved]` get the requested type from? I've tried looking into the docs on the attribute, but couldn't quite get my head around it.

Any information about that would be much appreciated!

bdach commented:
@sh0ckR6 please read [this primer article](https://github.com/ppy/osu-framework/wiki/Dependency-Injection) on dependency injection basics as implemented in the framework.

sh0ckR6 commented:
Thanks for the link, didn't think to look in the framework wiki pages. I'll work on the changes when I get home.

sh0ckR6 committed:
Resolve addressed issues

+ Stopped using testing methods in non-testing classes
+ Resolve Player and add OnSeek event
+ Take bindings away from BarHitErrorMeter
+ Add support for ColourHitErrorMeter

sh0ckR6 commented:
I've updated the PR with the requested changes (and added support for `ColourHitErrorMeter` along the way!)

I'll also add tests for the new changes, but it might take me a while. Unsure with how the visual testing system works, so I need to read about it on the wiki before getting started.

bdach commented on .idea/.idea.osu.Desktop/.idea/discord.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="DiscordProjectSettings">
+    <option name="show" value="ASK" />
+    <option name="description" value="" />
+  </component>
+</project>
```
should not be committed. please remove

sh0ckR6 replied:
Yeah, I'd realized right after pushing that this slipped into the commit. Didn't want to push this change until I received a review to avoid making unnecessary CI builds.

bdach commented on osu.Game/Screens/Play/HUD/HitErrorMeters/HitErrorMeter.cs:
```diff
@@ -67,6 +73,8 @@ protected Color4 GetColourForHitResult(HitResult result)
             }
         }
 
+        public abstract void Clear();
```
needs xmldoc (explaining that this gets called on seek and what implementors should do about it)

peppy replied:
This only needs to be `protected`.

sh0ckR6 replied:
This will break tests (unless there's something I'm missing with how my tests are written).

peppy replied:
Fine to leave it public then, I guess.

bdach commented on osu.Game/Screens/Play/SongProgress.cs:
```diff
@@ -35,6 +35,8 @@ public class SongProgress : OverlayContainer, ISkinnableDrawable
         private readonly SongProgressGraph graph;
         private readonly SongProgressInfo info;
 
+        public SongProgressBar Bar => bar;
```
no longer used, should be removed

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -584,7 +586,11 @@ private void performUserRequestedSkip()
+        public void Seek(double time)
+        {
+            GameplayClockContainer.Seek(time);
+            OnSeek.Invoke();
```
should be null-checked

sh0ckR6 committed:
Remove discord.xml

Not sure how that slipped in there, but it's gone now!

sh0ckR6 committed:
Remove unused property in `SongProgress`

sh0ckR6 committed:
Add xmldoc to HitErrorMeter.Clear

Explains how the method is called and what inheritors should do when implementing it.

sh0ckR6 committed:
Add null-check when invoking OnSeek

sh0ckR6 committed:
Add tests for clearing `HitErrorMeter`

Works with both `BarHitErrorMeter` and `ColourHitErrorMeter`

sh0ckR6 commented:
PR has been updated with tests for `HitErrorMeter`

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/HitErrorMeter.cs:
```diff
@@ -34,6 +37,9 @@ protected override void LoadComplete()
             base.LoadComplete();
 
+            if (player != null)
+                player.OnSeek += Clear;
```
This is a reference memory leak. For instance, if switching skins or using the skin layout editor, references to each of the `HitErrorMeter`s created along the way are going to be held by `player.OnSeek` events.

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs:
```diff
@@ -279,5 +279,7 @@ protected override void LoadComplete()
             }
         }
+
+        public override void Clear() => judgementsContainer.Clear(true);
```
```suggestion
        public override void Clear() => judgementsContainer.Clear();
```

disposal argument not required; it is the default.

peppy commented on osu.Game/Screens/Play/HUD/HitErrorMeters/ColourHitErrorMeter.cs:
```diff
@@ -33,6 +33,8 @@ protected override void OnNewJudgement(JudgementResult judgement)
             judgementsFlow.Push(GetColourForHitResult(judgement.Type));
         }
 
+        public override void Clear() => judgementsFlow.Clear(true);
```
```suggestion
        public override void Clear() => judgementsFlow.Clear();
```
As above.

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -69,6 +69,8 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
 
         public Action RestartRequested;
 
+        public Action OnSeek;
```
No xmldoc. Also not marked as an `event` which means anything could overwrite the previous usage.

peppy replied:
In addition, I'd move this inside `GameplayClockContainer`, as `Player` does not seem like the right place to be handling / binding to seek operations.

sh0ckR6 replied:
> No xmldoc. Also not marked as an `event` which means anything could overwrite the previous usage.

Resolved.

> In addition, I'd move this inside `GameplayClockContainer`, as Player does not seem like the right place to be handling / binding to seek operations.

With this change, all of `GameplayClockContainer` is now cached, not just its `GameplayClock`.

sh0ckR6 committed:
Address most issues

sh0ckR6 committed:
Resolve failed test compilation

sh0ckR6 committed:
Resolve `GameplayClockContainer` instead of `Player`

bdach reviewed (dismissed):
seems fine

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -584,7 +584,10 @@ private void performUserRequestedSkip()
         /// </summary>
         /// <param name="time">The destination time to seek to.</param>
-        public void Seek(double time) => GameplayClockContainer.Seek(time);
+        public void Seek(double time)
+        {
+            GameplayClockContainer.Seek(time);
+        }
```
```suggestion
        public void Seek(double time) => GameplayClockContainer.Seek(time);
```

Minor but can be reverted.

sh0ckR6 replied:
Done!

sh0ckR6 committed:
Fully revert `Player`

peppy reviewed (approved):
Seems good

bdach committed:
Merge branch 'master' into 14470-ur-not-updating
